packages/astro/src/core/middleware/callMiddleware.ts

```diff
@@ -42,54 +42,37 @@ export async function callMiddleware(
 ): Promise<Response> {
 	let nextCalled = false;
 	let responseFunctionPromise: Promise<Response> | Response | undefined = undefined;
-	const next: MiddlewareNext = async () => {
+
+	const response = await onRequest(apiContext, async () => {
 		nextCalled = true;
 		responseFunctionPromise = responseFunction();
 		return responseFunctionPromise;
-	};
+	});
 
-	let middlewarePromise = onRequest(apiContext, next);
+	if (response && response instanceof Response) {
+		return response;
+	}
 
-	return await Promise.resolve(middlewarePromise).then(async (value) => {
-		// first we check if `next` was called
+	if (response === undefined) {
+		/**
+		 * This seems unintentional, but it might be breaking to backtrack.
```

Comment:
This comment is very cryptic, I am not sure about its meaning. It might be a comment for yourself, but other people need to understand too. Could you provide more explanation?

Reply:
A middleware returning undefined is not a documented pattern. Here, we attempt to correct that misuse by directly using the response returned by renderPage/renderEndpoint. That seems unintentional.

Reply:
That's probably a leftover of the initial logic. Initially, we designed the feature in a way where a user could call just `next()` without returning anything from the middleware.

We changed that behaviour to be more in line with how other middleware in the JS ecosystem works. So yeah, we can remove that check

Reply:
That makes sense. Now, all that `callMiddleware()` needs to do is throw `MiddlewareNotAResponse`. That seems simple enough that it can be done directly, instead being extracted out into a function. What do you think?

Reply:
Yeah that sounds good to me!